src/itemdef.cpp

```diff
@@ -400,13 +411,33 @@ class CItemDefManager: public IWritableItemDefManager
 		return cc->inventory_texture;
 	}
 	// Get item wield mesh
-	virtual scene::IMesh* getWieldMesh(const std::string &name,
+	virtual ItemMesh* getWieldMesh(const std::string &name,
 			Client *client) const
 	{
 		ClientCached *cc = getClientCached(name, client);
 		if(!cc)
 			return NULL;
-		return cc->wield_mesh;
+		return &(cc->wield_mesh);
+	}
+
+	// Get item palette
+	virtual Palette* getPalette(const std::string &name,
+			Client *client) const
+	{
+		ClientCached *cc = getClientCached(name, client);
+		if(!cc)
+			return NULL;
+		return cc->palette;
+	}
+
+	virtual video::SColor getItemstackColor(const ItemStack &stack,
+		Client *client) const
+	{
+		Palette *palette = getPalette(stack.name, client);
+		const std::string &index = stack.metadata.getString("paletteindex", 0);
+		bool use_palette = (palette != NULL) && (index != "");
+		return (use_palette) ?
+			(*palette)[mystoi(index, 0, 255)] : get(stack.name).color;
```

Comment:
I chose palettes because they are flexible and can be overridden by texture packs. But if custom colors are also requested, I can do it!